src/test/java/com/snowflake/kafka/connector/internal/SinkServiceIT.java

```diff
@@ -1147,4 +1148,229 @@ public void testCleanerRecoverListCount() throws Exception {
+            .addTask(table, topic, partition)
+            .build();
+    offset = 0;
+    // Create sink record with same offset
+    SinkRecord record2 =
+        new SinkRecord(
+            topic, partition, Schema.STRING_SCHEMA, "test", input.schema(), input.value(), offset);
+    service2.insert(record2);
+    // The record should not be inserted in the buffer
+    Assert.assertTrue(
+        ((SnowflakeSinkServiceV1) service2)
+            .isPartitionBufferEmpty(SnowflakeSinkServiceV1.getNameIndex(topic, partition)));
+  }
+
+  @Ignore
+  @Test(expected = SnowflakeKafkaConnectorException.class)
+  // files write to FDB for an older clientSequencer should be discarded
+  public void testFileDiscardWithExactlyOnceSemantics() throws Exception {
+    conn.createTable(table);
+    conn.createStage(stage);
+    SnowflakeSinkService service =
+        SnowflakeSinkServiceFactory.builder(conn)
+            .setRecordNumber(1)
+            .setDeliveryGuarantee(
+                SnowflakeSinkConnectorConfig.IngestionDeliveryGuarantee.EXACTLY_ONCE)
+            .addTask(table, topic, partition)
+            .build();
+
+    SnowflakeConverter converter = new SnowflakeJsonConverter();
+    SchemaAndValue input =
+        converter.toConnectData(topic, "{\"name\":\"test\"}".getBytes(StandardCharsets.UTF_8));
+
+    SnowflakeIngestionService ingestionService = conn.buildIngestService(stage, pipe);
+
+    long offset = 0;
+    // Create sink record
+    SinkRecord record1 =
+        new SinkRecord(
+            topic, partition, Schema.STRING_SCHEMA, "test", input.schema(), input.value(), offset);
+
+    service.insert(record1);
+    // Confirm the record is flushed to stage
+    TestUtils.assertWithRetry(
+        () ->
+            conn.listStage(
+                        stage,
+                        FileNameUtils.filePrefix(TestUtils.TEST_CONNECTOR_NAME, table, partition))
+                    .size()
+                == 1,
+        5,
+        4);
+
+    // Make sure correct client sequencer
+    ClientStatusResponse clientStatusResponse = ingestionService.getClientStatus();
+    assert clientStatusResponse.getClientSequencer() == 0;
+    assert clientStatusResponse.getOffsetToken() == null;
+
+    // initialize a new sink service without closing the old service yet
+    SnowflakeSinkService service2 =
+        SnowflakeSinkServiceFactory.builder(conn)
+            .setRecordNumber(1)
+            .setDeliveryGuarantee(
+                SnowflakeSinkConnectorConfig.IngestionDeliveryGuarantee.EXACTLY_ONCE)
+            .addTask(table, topic, partition)
+            .build();
+    offset = 1;
+    // Create sink record with same offset
+    SinkRecord record2 =
+        new SinkRecord(
+            topic, partition, Schema.STRING_SCHEMA, "test", input.schema(), input.value(), offset);
+    // Should update the client sequencer
+    service2.insert(record2);
+    // call snowpipe ingest api
+    service.callAllGetOffset();
```

Comment:
so this will throw an exception? Can you catch the exception and assert the exception type? I think we discussed that it will throw IngestException so may be we can just catch it? 

And then may be we can call  `service2.callAllGetOffset();` to see if offset 1 is getting ingested? 

In an ideal world we would expect the offset 0 to be ingested by `service2` (The newly woken up instance after rebalance) since precommit failed in `service0` and offset 0 was never ingested in SF.

Reply:
I ran this test. we should catch exception and assert. 
I see all retries with 
```Message: Pipe has older version of clientSequencer. Please try again with latest clientSequencer.,
Data: null
}```
Which seems corrcet and behavior also seemed to be correct.

i.e on this exception technically there is another node with higher sequencer trying to ingest the data. so this instance of connector should stop working. i.e stop abruptly.